### DISCUSSION TITLE
Should the application controller not be distributing the burden across its replicas?

### DISCUSSION THREAD

**Discussion by djfinnoy** [argoproj/argo-cd#12808] [General]
We are running `v2.6.2`.
Noticed that sync events affecting at least 10 or so applications is causing sluggish sync operations, and crashing the `argocd-application-controller-0` pod.

```
$ kubectl top po -n argocd
NAME                                                CPU(cores)   MEMORY(bytes)   
argocd-application-controller-0                     707m         381Mi           
argocd-application-controller-1                     1m           44Mi            
argocd-application-controller-2                     2m           49Mi             
argocd-redis-ha-haproxy-6c646f7788-8nbfq            6m           70Mi            
argocd-redis-ha-haproxy-6c646f7788-97j6t            4m           70Mi            
argocd-redis-ha-haproxy-6c646f7788-z2kbq            5m           70Mi            
argocd-redis-ha-server-0                            103m         39Mi            
argocd-redis-ha-server-1                            140m         39Mi            
argocd-redis-ha-server-2                            99m          39Mi            
argocd-repo-server-75dd445558-4dqvv                 4m           65Mi            
argocd-repo-server-75dd445558-79xqs                 4m           78Mi            
argocd-repo-server-75dd445558-b56cm                 4m           145Mi           
argocd-repo-server-75dd445558-gcngp                 4m           147Mi           
argocd-repo-server-75dd445558-mccl2                 4m           96Mi            
```

```
$ kubectl get pods

NAME                                                READY   STATUS             RESTARTS        AGE
argocd-application-controller-0                     0/1     CrashLoopBackOff   2 (9s ago)      3m22s
argocd-application-controller-1                     1/1     Running            0               27d
argocd-application-controller-2                     1/1     Running            0               27d

```

To me, it seems like only the first pod of the application controller is being tasked with implementing syncs / application changes, and that it is crashing from running out of CPU resources.
Should it not be distributing the burden across all three replicas?
Is this some side effect of running app-of-apps patterns?

Here is a snippet of my application-controller config:

```
    controller:
      logLevel: error
      enableStatefulSet: true
      replicas: 3
      resources:
        limits:
          cpu: 750m
          memory: 2Gi
        requests:
          cpu: 250m
          memory: 512Mi
```

**Comment by jannfis** [comment]
Hi @djfinnoy . I'm not very knowledgeable about how the Helm chart configures Argo CD under the hood, but there are are a couple of factors to consider when scaling up the application controller workload.

* First, read https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller
* Then, there's also (at least one) open bug(s) with sharding: https://github.com/argoproj/argo-cd/issues/9633
* And a couple of ideas to improve: https://github.com/argoproj/argo-cd/pull/8340 (which we probably will cut into multiple separate pieces and evaluate/implement progressively)

HTH.

**Reply by djfinnoy** [reply]
Thank you for the information, I'll have a read.

**Comment by djfinnoy** [comment]
So based on what I read in @jannfis' response, application-controller sharding happens on a per-cluster basis.
At the moment, we run one ArgoCD instance per cluster, so it won't be possible for us to utilize application-controller sharding to improve performance. Is this correct?

**Reply by jannfis** [reply]
Yes. You can, however, scale the application controller vertically:

- Give it more CPU
- Increase operation and status processors as described in the document I linked

Sharding on application level is something we are considering, but have not yet come to a final conclusion or consensus yet.
